qiskit/transpiler/passes/synthesis/hls_plugins.py

```diff
@@ -0,0 +1,924 @@
+            # Unfortunately we occasionally have custom instructions called "mcx"
+            # which get wrongly caught by the plugin interface. A simple solution is
+            # to return None in this case, since HLS would proceed to examine
+            # their definition as it should.
+            return None
+
+        num_ctrl_qubits = high_level_object.num_ctrl_qubits
+        decomposition = synth_mcx_gray_code(num_ctrl_qubits)
+        return decomposition
+
+
+class MCXSynthesisMCPhase(HighLevelSynthesisPlugin):
+    r"""Synthesis plugin for a multi-controlled X gate based on the
+    implementation for MCPhaseGate.
+
+    This plugin name is :``mcx.mcphase`` which can be used as the key on
+    an :class:`~.HLSConfig` object to use this method with :class:`~.HighLevelSynthesis`.
+
+    For a multi-controlled X gate with :math:`k` control qubits this synthesis
+    method requires no additional clean auxiliary qubits. The synthesized
+    circuit consists of :math:`k + 1` qubits.
+    """
+
+    def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **options):
+        """Run synthesis for the given MCX gate."""
+
+        if high_level_object.name != "mcx":
+            raise TranspilerError("MCXSynthesisMCPhase only accepts objects of type MCXGate")
+
+        if not isinstance(high_level_object, (MCXGate, C3XGate, C4XGate)):
+            # Unfortunately we occasionally have custom instructions called "mcx"
+            # which get wrongly caught by the plugin interface. A simple solution is
+            # to return None in this case, since HLS would proceed to examine
+            # their definition as it should.
+            return None
+
+        num_ctrl_qubits = high_level_object.num_ctrl_qubits
+        decomposition = synth_mcx_mcphase(num_ctrl_qubits)
+        return decomposition
+
+
+class MCXSynthesisDefault(HighLevelSynthesisPlugin):
+    r"""The default synthesis plugin for a multi-controlled X gate.
+
+    This plugin name is :``mcx.default`` which can be used as the key on
+    an :class:`~.HLSConfig` object to use this method with :class:`~.HighLevelSynthesis`.
+    """
+
+    def run(self, high_level_object, coupling_map=None, target=None, qubits=None, **options):
+        """Run synthesis for the given MCX gate."""
+
+        if not isinstance(high_level_object, (MCXGate, C3XGate, C4XGate)):
+            # Unfortunately we occasionally have custom instructions called "mcx"
+            # which get wrongly caught by the plugin interface. A simple solution is
+            # to return None in this case, since HLS would proceed to examine
+            # their definition as it should.
+            return None
+
+        # Iteratively run other synthesis methods available
+
+        if (
+            decomposition := MCXSynthesisNCleanM15().run(
+                high_level_object, coupling_map, target, qubits, **options
+            )
+        ) is not None:
+            return decomposition
```

Comment:
This works and we can probably do it this way for the beginning, but in future we'll surely need a method to query how many auxiliary qubits a plugin requires. We could already add it here and have a code structure that directly selects the correct plugin based on the number of available auxiliary qubits. What do you think?

Reply:
As discussed, I have also marked the `MCXGate`'s function `get_num_ancilla_qubits` for pending deprecation. 

IMHO, in general a plugin should not know how many ancilla qubits it would require before it runs: depending on the high-level object itself, on options, on luck, or on random seed, it may return different decompositions using different numbers of ancilla qubits. (But note that a plugin does receive how many ancilla qubits are available, and cannot use more than this number of qubits).

Reply:
I agree for plugins that can use different strategies, but the specific MCX plugins e.g. `MCXSynthesisNCleanM15` have exactly one synthesis method that they use, so it would be safe to add it there, no? Also, there's no way to infer the number of ancilla qubits otherwise (without actually running the decomposition), if we deprecate it on `MCXGate`.

Reply:
We discussed this offline, and decided to leave it as is implemented right now. We will need to rethink this in the future once we start looking at analysis passes that "optimally" divide the ancilla qubits between different high-level-objects.